gslib/context_config.py

```diff
@@ -107,6 +113,55 @@ def _SplitPemIntoSections(contents, logger):
   return result
 
 
+def _check_path():
+  """Checks for context aware metadata. If it exists, returns the absolute path;
+  otherwise returns None.
+  Returns:
+      str: absolute path if exists and None otherwise.
+  """
+  metadata_path = os.path.expanduser(_CONTEXT_AWARE_METADATA_PATH)
+  if not os.path.exists(metadata_path):
+    return None
+  return metadata_path
+
+
+def _read_metadata_file(metadata_path):
+  """Loads context aware metadata from the given path.
+  Returns:
+      Dict[str, str]: The metadata.
+  Raises:
+      CertProvisionError: If failed to parse metadata as JSON.
+  """
+  try:
+    with open(metadata_path) as f:
+      metadata = json.load(f)
+  except ValueError as e:
+    raise CertProvisionError(e)
+  return metadata
+
+
+def _default_command():
+  """Loads default cert provider command.
+  Returns:
+      str: The default command.
+  Raises:
+      CertProvisionError: If command cannot be found.
+  """
+  metadata_path = _check_path()
+  if metadata_path:
+    metadata_json = _read_metadata_file(metadata_path)
+
+    if _CERT_PROVIDER_COMMAND not in metadata_json:
+      raise CertProvisionError("Client certificate provider is not found.")
+
+    command = metadata_json[_CERT_PROVIDER_COMMAND]
+    if (_CERT_PROVIDER_COMMAND_PASSPHRASE_OPTION not in command):
+      command.append(_CERT_PROVIDER_COMMAND_PASSPHRASE_OPTION)
```

Comment:
Wouldn't this be appending to a string, based on http://go/aip/auth/4114 ?

You can only append to lists in Python. Strings use `+` for concatenation.

Reply:
The '--with-passphrase' in this case for security purposes.

https://github.com/googleapis/google-auth-library-python/blob/e6976879b392508c022610ab3ea2ea55c7089c63/google/auth/transport/_mtls_helper.py#L176

Reply:
Please explain this better.

First, I don't understand how you're using `append` between two strings. That should error:
```
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 'abc'.append('123')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'str' object has no attribute 'append'
```

Second, based on this line of the auth/4114, it looks like `cert_provider_command` is any user inputted command:
```
"cert_provider_command": "[absolute_path_to_provider_command] --fetch_client_cert"
```

How do we know the command will support "--with-passphrase"? The proposal doesn't require it: http://go/aip/auth/4114

Reply:
1. The field 'cert_provider_command' is a list of strings, so I think `json.load(..)` returns a list in this case.
2. The metadata file and its associated command is Google provided so this '--with-passphrase' assumption should be safe to make.

Reply:
I can make a change to the AIP to indicate that `--with-passphrase` should be supported by the command.

Reply:
1) Maybe the AIP should also be updated to show a list of strings for "cert_provider_command"? It looks like a single string right now under "Obtaining the Default Device Certificate". The flag "--fetch_client_cert" is also unexplained.

2) Gotcha.